src/nnpack/nnlib.jl

```diff
@@ -0,0 +1,86 @@
+end
+
+function softmax!(y::AbstractVecOrMat{Float32},x::AbstractVecOrMat{Float32}; inplace::Bool = true, nthreads = NNPACK_CPU_THREADS)
+    threadpool = nthreads != NNPACK_CPU_THREADS ? pthreadpool_create(nthreads) : shared_threadpool
+    nnp_softmax_output(x, y, threadpool = threadpool)
+end
+
+function softmax(x::AbstractVecOrMat{Float32}; nthreads = NNPACK_CPU_THREADS)
+    threadpool = nthreads != NNPACK_CPU_THREADS ? pthreadpool_create(nthreads) : shared_threadpool
+    nnp_softmax_output(x, similar(x), threadpool = threadpool)
+end
+
+function maxpool(x::AA{4}, k; pad = map(_->0,k), stride = k, nthreads = NNPACK_CPU_THREADS)
+    threadpool = nthreads != NNPACK_CPU_THREADS ? pthreadpool_create(nthreads) : shared_threadpool
+    pad_, stride_ = expand(Val{length(k)}, pad), expand(Val{length(k)}, stride)
+    ((size(x, 1) - k[1] + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - k[2] + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
+    maxpool!(similar(x, pdims(size(x), k, pad_, stride_)), x, k, pad = pad_, stride = stride_, threadpool = threadpool)
+end
+
+maxpool!(y::AA{4}, x::AA{4}, k; pad = map(_->0,k), stride = k, threadpool = nothing) =
+    nnp_max_pooling_output(x, y, k, padding = expand(Val{length(k)}, pad), stride = expand(Val{length(k)}, stride), threadpool = threadpool)
+
+function conv(x::AA{4}, w::AA{4}; pad = 0, stride = 1, dilation = 1, algo = UInt32(0), nthreads = NNPACK_CPU_THREADS)
+    threadpool = nthreads != NNPACK_CPU_THREADS ? pthreadpool_create(nthreads) : shared_threadpool
+    dilation == 1 || dilation == (1, 1) || error("NNPACK does not support dilation > 1")
+    pad_, stride_ = padtuple(x, pad), padtuple(x, stride)
+    ((size(x, 1) - size(w, 1) + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - size(w, 2) + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
+    w = reverse(reverse(w, dims=1), dims=2) # Needs to be fixed once the flipkernel and crosscor PR is merged
+    y = similar(x, cdims(size(x), dilation_dims(w, dilation), pad_, stride_))
+    b = zeros(Float32, size(y, 3))
+    conv!(y, x, w, b, pad = pad_, stride = stride_, dilation = dilation, algo = UInt32(algo), threadpool = threadpool)
+end
+
+function conv(x::AA{4}, w::AA{4}, b::AA{1}; pad = 0, stride = 1, dilation = 1, algo = UInt32(0), nthreads = NNPACK_CPU_THREADS)
+    threadpool = nthreads != NNPACK_CPU_THREADS ? pthreadpool_create(nthreads) : shared_threadpool
+    dilation == 1 || dilation == (1, 1) || error("NNPACK does not support dilation > 1")
+    pad_, stride_ = padtuple(x, pad), padtuple(x, stride)
+    ((size(x, 1) - size(w, 1) + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - size(w, 2) + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
+    w = reverse(reverse(w, dims=1), dims=2) # Needs to be fixed once the flipkernel and crosscor PR is merged
```

Comment:
Should be possible to combine this into one with:
```suggestion
    w = reverse(w, dims=(1,2)) # Needs to be fixed once the flipkernel and crosscor PR is merged
```

Reply:
`reverse` function is expecting an integer as dims. Its not taking a tuple